File: examples/ts/auction/Auction.test.ts
import { describe, it, expect } from 'vitest';
import { readFileSync } from 'node:fs';
import { join, dirname } from 'node:path';
import { fileURLToPath } from 'node:url';
import { TestContract } from 'runar-testing';

const __dirname = dirname(fileURLToPath(import.meta.url));
const source = readFileSync(join(__dirname, 'Auction.runar.ts'), 'utf8');

const AUCTIONEER = '02' + 'aa'.repeat(32);
const BIDDER_A = '02' + 'bb'.repeat(32);
const BIDDER_B = '02' + 'cc'.repeat(32);
const MOCK_SIG = '30' + 'ff'.repeat(35);
const DEADLINE = 500000n;

describe('Auction', () => {
  function makeAuction(highestBid = 0n) {
    const auction = TestContract.fromSource(source, {
      auctioneer: AUCTIONEER,
      highestBidder: BIDDER_A,
      highestBid,
      deadline: DEADLINE,
    });
    // Set locktime before deadline (bidding is open)
    auction.setMockPreimage({ locktime: DEADLINE - 1n });
    return auction;
  }

  it('accepts a valid bid above current highest', () => {
    const auction = makeAuction(100n);
    const result = auction.call('bid', { sig: MOCK_SIG, bidder: BIDDER_B, bidAmount: 200n });
    expect(result.success).toBe(true);
    expect(auction.state.highestBid).toBe(200n);
    expect(auction.state.highestBidder).toBe(BIDDER_B);
  });

  it('rejects a bid below current highest', () => {
    const auction = makeAuction(100n);
    const result = auction.call('bid', { sig: MOCK_SIG, bidder: BIDDER_B, bidAmount: 50n });
    expect(result.success).toBe(false);
  });

  it('rejects a bid after deadline', () => {
    const auction = makeAuction(100n);
    auction.setMockPreimage({ locktime: DEADLINE + 1n });
    const result = auction.call('bid', { sig: MOCK_SIG, bidder: BIDDER_B, bidAmount: 200n });
    expect(result.success).toBe(false);
  });

  it('allows close after deadline', () => {
    const auction = makeAuction(100n);
    auction.setMockPreimage({ locktime: DEADLINE });
    const result = auction.call('close', { sig: MOCK_SIG });
    expect(result.success).toBe(true);
  });

  it('rejects close before deadline', () => {
    const auction = makeAuction(100n);
    auction.setMockPreimage({ locktime: DEADLINE - 1n });
    const result = auction.call('close', { sig: MOCK_SIG });
    expect(result.success).toBe(false);
  });

  it('tracks multiple bids in sequence', () => {
    const auction = makeAuction(0n);

    auction.call('bid', { sig: MOCK_SIG, bidder: BIDDER_A, bidAmount: 100n });
    expect(auction.state.highestBid).toBe(100n);

    auction.call('bid', { sig: MOCK_SIG, bidder: BIDDER_B, bidAmount: 200n });
    expect(auction.state.highestBid).toBe(200n);
    expect(auction.state.highestBidder).toBe(BIDDER_B);
  });
});

File: examples/zig/auction/Auction_test.zig
const std = @import("std");

const root = @import("../examples_test.zig");
const runar = @import("runar");
const Auction = @import("Auction.runar.zig").Auction;

const contract_source = @embedFile("Auction.runar.zig");

test "compile-check Auction.runar.zig" {
    const allocator = std.testing.allocator;
    var tmp = std.testing.tmpDir(.{});
    defer tmp.cleanup();
    try tmp.dir.writeFile(.{ .sub_path = "Auction.runar.zig", .data = contract_source });

    const path = try std.fs.path.join(allocator, &.{ ".zig-cache", "tmp", tmp.sub_path[0..], "Auction.runar.zig" });
    defer allocator.free(path);
    try root.runar.compileCheckFile(allocator, path);
    try root.runar.compileCheckSource(allocator, contract_source, "Auction.runar.zig");
}

test "auction init stores constructor fields" {
    const auction = Auction.init(runar.ALICE.pubKey, runar.BOB.pubKey, 100, 500);

    try std.testing.expectEqualSlices(u8, runar.ALICE.pubKey, auction.auctioneer);
    try std.testing.expectEqualSlices(u8, runar.BOB.pubKey, auction.highestBidder);
    try std.testing.expectEqual(@as(i64, 100), auction.highestBid);
    try std.testing.expectEqual(@as(i64, 500), auction.deadline);
}

test "auction accepts a higher bid before deadline through the real contract" {
    var runtime = runar.StatefulSmartContract.init(std.testing.allocator);
    defer runtime.deinit();
    var auction = Auction.init(runar.ALICE.pubKey, runar.ALICE.pubKey, 100, 500);
    const ctx = try runar.StatefulContext.init(&runtime, runar.mockPreimage(.{ .locktime = 499 }));

    auction.bid(ctx, runar.signTestMessage(runar.BOB), runar.BOB.pubKey, 150);
    try std.testing.expectEqualSlices(u8, runar.BOB.pubKey, auction.highestBidder);
    try std.testing.expectEqual(@as(i64, 150), auction.highestBid);
}

test "auction closes at or after the deadline through the real contract" {
    var runtime = runar.StatefulSmartContract.init(std.testing.allocator);
    defer runtime.deinit();
    const auction = Auction.init(runar.ALICE.pubKey, runar.BOB.pubKey, 150, 500);
    const ctx = try runar.StatefulContext.init(&runtime, runar.mockPreimage(.{ .locktime = 500 }));

    auction.close(ctx, runar.signTestMessage(runar.ALICE));
}

test "auction rejects low or late bids" {
    try root.expectAssertFailure("auction-bid-too-low");
    try root.expectAssertFailure("auction-bid-too-late");
}

test "auction rejects early close and wrong closing signature" {
    try root.expectAssertFailure("auction-close-too-early");
    try root.expectAssertFailure("auction-close-wrong-sig");
}
